File: web/giga/tokenizer.js
// Превращение номеров, которые выдаёт модель, обратно в текст.
//
// Модель говорит кусочками слов (SentencePiece). Чтобы собрать из них текст,
// нужен только список этих кусочков — он лежит в v3_e2e_rnnt_tokenizer.model.
// Файл записан в формате protobuf, но нам нужно одно поле, поэтому разбираем
// его сами, как и swift/Tokenizer.swift:
//
//   ModelProto    { repeated SentencePiece pieces = 1 }
//   SentencePiece { optional string piece = 1 }

const utf8 = new TextDecoder("utf-8");

function varint(d, from) {
  let result = 0n, shift = 0n, i = from;
  while (i < d.length) {
    const b = d[i++];
    result |= BigInt(b & 0x7f) << shift;
    if ((b & 0x80) === 0) return [Number(result), i];
    shift += 7n;
    if (shift > 63n) return null;
  }
  return null;
}

function skip(d, from, wire) {
  switch (wire) {
    case 0: { const v = varint(d, from); return v ? v[1] : null; }
    case 1: return from + 8 <= d.length ? from + 8 : null;
    case 2: {
      const v = varint(d, from);
      if (!v) return null;
      const end = v[1] + v[0];
      return end <= d.length ? end : null;
    }
    case 5: return from + 4 <= d.length ? from + 4 : null;
    default: return null;
  }
}

/** Внутри SentencePiece берём первое поле — саму строку кусочка. */
function piece(d, from, to) {
  let i = from;
  while (i < to) {
    const t = varint(d, i);
    if (!t) break;
    i = t[1];
    const field = Math.floor(t[0] / 8), wire = t[0] & 7;
    if (field === 1 && wire === 2) {
      const l = varint(d, i);
      if (!l) break;
      const end = Math.min(l[1] + l[0], to);
      return utf8.decode(d.subarray(l[1], end));
    }
    const next = skip(d, i, wire);
    if (next === null) break;
    i = next;
  }
  return "";
}

export class Tokenizer {
  /** bytes — содержимое файла *_tokenizer.model (Uint8Array или ArrayBuffer). */
  constructor(bytes) {
    const d = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
    const out = [];
    let i = 0;
    while (i < d.length) {
      const t = varint(d, i);
      if (!t) break;
      i = t[1];
      const field = Math.floor(t[0] / 8), wire = t[0] & 7;
      if (field === 1 && wire === 2) {
        const l = varint(d, i);
        if (!l) break;
        const end = l[1] + l[0];
        if (end > d.length) break;
        out.push(piece(d, l[1], end));
        i = end;
      } else {
        const next = skip(d, i, wire);
        if (next === null) break;
        i = next;
      }
    }
    if (out.length === 0) throw new Error("не разобрал токенизатор");
    this.pieces = out;
  }

  /** Номер «пустышки» — модель выдаёт его, когда сказать нечего. */
  get blankId() {
    return this.pieces.length;
  }

  /** Собирает текст: кусочки склеиваются, знак ▁ означает пробел перед словом. */
  decode(ids) {
    let s = "";
    for (const id of ids) if (id >= 0 && id < this.pieces.length) s += this.pieces[id];
    return s.replaceAll("▁", " ").trim();
  }
}

File: windows/gigapisar/asr/tokenizer.go
package asr

import (
	"errors"
	"strings"
)

// Tokenizer — кусочки слов из *_tokenizer.model (protobuf SentencePiece):
// ModelProto { repeated SentencePiece pieces = 1 }, SentencePiece { string piece = 1 }.
type Tokenizer struct{ Pieces []string }

func varint(d []byte, i int) (uint64, int, bool) {
	var r uint64
	shift := 0
	for i < len(d) {
		b := d[i]
		i++
		r |= uint64(b&0x7f) << shift
		if b&0x80 == 0 {
			return r, i, true
		}
		shift += 7
		if shift > 63 {
			return 0, i, false
		}
	}
	return 0, i, false
}

func skipField(d []byte, i int, wire int) (int, bool) {
	switch wire {
	case 0:
		_, j, ok := varint(d, i)
		return j, ok
	case 1:
		return i + 8, i+8 <= len(d)
	case 2:
		l, j, ok := varint(d, i)
		if !ok || j+int(l) > len(d) {
			return 0, false
		}
		return j + int(l), true
	case 5:
		return i + 4, i+4 <= len(d)
	}
	return 0, false
}

func piece(d []byte) string {
	i := 0
	for i < len(d) {
		t, j, ok := varint(d, i)
		if !ok {
			break
		}
		i = j
		if t>>3 == 1 && t&7 == 2 {
			l, j, ok := varint(d, i)
			if !ok {
				break
			}
			end := j + int(l)
			if end > len(d) {
				end = len(d)
			}
			return string(d[j:end])
		}
		i, ok = skipField(d, i, int(t&7))
		if !ok {
			break
		}
	}
	return ""
}

func NewTokenizer(bytes []byte) (*Tokenizer, error) {
	var out []string
	i := 0
	for i < len(bytes) {
		t, j, ok := varint(bytes, i)
		if !ok {
			break
		}
		i = j
		if t>>3 == 1 && t&7 == 2 {
			l, j, ok := varint(bytes, i)
			if !ok || j+int(l) > len(bytes) {
				break
			}
			out = append(out, piece(bytes[j:j+int(l)]))
			i = j + int(l)
		} else {
			i, ok = skipField(bytes, i, int(t&7))
			if !ok {
				break
			}
		}
	}
	if len(out) == 0 {
		return nil, errors.New("не разобрал токенизатор")
	}
	return &Tokenizer{Pieces: out}, nil
}

func (t *Tokenizer) BlankID() int { return len(t.Pieces) }

func (t *Tokenizer) Decode(ids []int) string {
	var sb strings.Builder
	for _, id := range ids {
		if id >= 0 && id < len(t.Pieces) {
			sb.WriteString(t.Pieces[id])
		}
	}
	return strings.TrimSpace(strings.ReplaceAll(sb.String(), "▁", " "))
}
